google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java

```diff
@@ -442,22 +591,163 @@
+      final DocumentReference documentReference,
+      final OperationType operationType,
+      final BulkWriterOperationCallback operationCallback) {
+    return ApiFutures.transformAsync(
+        ApiFutures.immediateFuture(null),
+        new ApiAsyncFunction<Object, WriteResult>() {
+          public ApiFuture<WriteResult> apply(Object o) throws Exception {
+            return enqueueWriteHelper(documentReference, operationType, operationCallback, 0);
+          }
+        },
+        bulkWriterExecutor);
+  }
+
+  /**
+   * Adds the write to the appropriate batch queue and performs the bulkCommit. This helper method
+   * also catches failures and applies the user error callback and retrying if necessary.
+   */
+  private ApiFuture<WriteResult> enqueueWriteHelper(
+      final DocumentReference documentReference,
+      final OperationType operationType,
+      final BulkWriterOperationCallback operationCallback,
+      final int failedAttempts) {
+    List<BulkCommitBatch> operationBatchQueue = failedAttempts > 0 ? retryBatchQueue : batchQueue;
+    BulkCommitBatch bulkCommitBatch = getEligibleBatch(documentReference, operationBatchQueue);
+
+    // Send ready batches if this is the first attempt. Subsequent retry batches are scheduled
+    // after the initial batch returns.
+    if (failedAttempts == 0) {
+      sendReadyBatches(operationBatchQueue);
+    }
+
+    return ApiFutures.catchingAsync(
+        operationCallback.apply(bulkCommitBatch),
+        FirestoreException.class,
+        new ApiAsyncFunction<FirestoreException, WriteResult>() {
+          public ApiFuture<WriteResult> apply(FirestoreException exception)
+              throws BulkWriterError, ExecutionException, InterruptedException {
+            BulkWriterError bulkWriterError =
+                new BulkWriterError(
+                    exception.getStatus(),
+                    exception.getMessage(),
+                    documentReference,
+                    operationType,
+                    failedAttempts);
+
+            boolean shouldRetry = invokeUserErrorCallback(bulkWriterError).get();
+            logger.log(
+                Level.INFO,
+                String.format(
+                    "Running error callback on error code: %d, shouldRetry: %b",
+                    exception.getStatus().getCode().value(), shouldRetry));
+            if (!shouldRetry) {
+              throw bulkWriterError;
+            } else {
+              return enqueueWriteHelper(
+                  documentReference, operationType, operationCallback, failedAttempts + 1);
+            }
+          }
+        },
+        MoreExecutors.directExecutor());
+  }
+
+  /** Invokes the user error callback on the user callback executor and returns the result. */
+  private SettableApiFuture<Boolean> invokeUserErrorCallback(final BulkWriterError error) {
+    final SettableApiFuture<Boolean> callbackResult = SettableApiFuture.create();
+    errorExecutor.execute(
```

Comment:
If you make the default Exectuor the `bulkWriterExecutor`, it might be best to make the `errorExecutor` nullable. If no custom executor is set, you could invoke the `errorListener` directly.